ios/browser/debounce/debounce_service_factory.mm

```diff
@@ -0,0 +1,82 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/debounce/debounce_service_factory.h"
+
+#include "base/no_destructor.h"
+#include "brave/components/debounce/core/debounce_component_installer.h"
+#include "brave/components/debounce/core/debounce_service.h"
+#include "brave/ios/browser/api/debounce/debounce_service+private.h"
+#include "brave/ios/browser/application_context/brave_application_context_impl.h"
+#include "brave/ios/browser/debounce/debounce_service_factory+private.h"
+#include "brave/ios/browser/keyed_service/keyed_service_factory_wrapper+private.h"
+#include "components/keyed_service/core/keyed_service.h"
+#include "components/keyed_service/ios/browser_state_dependency_manager.h"
+#include "ios/chrome/browser/shared/model/application_context/application_context.h"
+#include "ios/chrome/browser/shared/model/browser_state/browser_state_otr_helper.h"
+#include "ios/chrome/browser/shared/model/browser_state/chrome_browser_state.h"
+#include "ios/web/public/browser_state.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+@implementation DebounceServiceFactory
++ (nullable id)serviceForBrowserState:(ChromeBrowserState*)browserState {
+  // Create and start the local data file service and component installer
+  debounce::DebounceService* debounceService =
+      debounce::DebounceServiceFactory::GetServiceForState(browserState);
+  return [[DebounceService alloc] initWithDebounceService:debounceService];
+}
+@end
+
+namespace debounce {
+
+// static
+debounce::DebounceService* DebounceServiceFactory::GetServiceForState(
+    ChromeBrowserState* browser_state) {
+  return static_cast<debounce::DebounceService*>(
+      GetInstance()->GetServiceForBrowserState(browser_state, true));
+}
+
+// static
+DebounceServiceFactory* DebounceServiceFactory::GetInstance() {
+  static base::NoDestructor<DebounceServiceFactory> instance;
+  return instance.get();
+}
+
+DebounceServiceFactory::DebounceServiceFactory()
+    : BrowserStateKeyedServiceFactory(
+          "DebounceService",
+          BrowserStateDependencyManager::GetInstance()) {}
+
+DebounceServiceFactory::~DebounceServiceFactory() = default;
+
+std::unique_ptr<KeyedService> DebounceServiceFactory::BuildServiceInstanceFor(
+    web::BrowserState* context) const {
+  BraveApplicationContextImpl* braveContext =
+      static_cast<BraveApplicationContextImpl*>(GetApplicationContext());
+  std::unique_ptr<debounce::DebounceService> service =
+      std::make_unique<debounce::DebounceService>(
+        braveContext->debounce_component_installer(),
+        braveContext->GetLocalState());
+        
+  return service;
+}
+
+bool DebounceServiceFactory::ServiceIsNULLWhileTesting() const {
+  return true;
+}
+
+bool DebounceServiceFactory::ServiceIsCreatedWithBrowserState() const {
+  return true;
```

Comment:
I'm not really sure. So yah I'm missing the `debounce_component_installer` call on `brave_application_context_impl` but it doesn't seem to matter. It works just fine. Shouldn't the call to  `debounce::DebounceServiceFactory::GetInstance()` on `brave_browser_state_keyed_service_factories` be enough? 

The other difference on with the two services is that the debounce service takes hold of the component installer, unlike url service which doesn't

Reply:
It works because you are eagerly creating the whole DebounceService which then creates the installer, I'm just pointing out that desktop doesn't do that. They only eagerly set up the installer, the service itself isn't eager.

`GetInstance` is not creating the service, it only sets up the factory

Reply:
Fair enough, I removed this eagerness and eagerly call `debounce_component_installer()`